File: CPP/ByRound/1479/1479_A_Searching_Local_Minimum.cpp
// Problem: CF 1479 A - Searching Local Minimum
// https://codeforces.com/contest/1479/problem/A

/*
 * Purpose:
 *   Find any local minimum in a permutation of integers 1 to n using at most 100 queries.
 *   A local minimum at index i satisfies a[i] < min(a[i-1], a[i+1]), with boundaries treated as +∞.
 *
 * Algorithm:
 *   Use binary search on the indices to locate a local minimum.
 *   At each step:
 *     - Query two adjacent elements: a[mid] and a[mid+1]
 *     - If a[mid] < a[mid+1], then the local minimum must be in [l, mid]
 *     - Otherwise, it must be in [mid+1, r]
 *   Continue until l == r, which points to a local minimum.
 *
 * Time Complexity: O(log n) - binary search over n elements with 2 queries per step.
 * Space Complexity: O(1) - only a few variables used, no additional space.
 */

#include <iostream>
using namespace std;

int main() {
  int n;
  cin >> n;
  int l = 1, r = n; // Initialize binary search range
  while (l < r) {
    int a, b, mid = (l + r) / 2; // Calculate mid-point
    cout << "? " << mid << endl;
    cin >> a; // Query value at mid
    cout << "? " << mid + 1 << endl;
    cin >> b; // Query value at mid+1
    if (a < b)
      r = mid; // Local min is in left half including mid
    else
      l = mid + 1; // Local min is in right half excluding mid
  }
  cout << "! " << l << "\n"; // Output the found local minimum index
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/A/1479_A_Searching_Local_Minimum.py
# Problem: CF 1479 A - Searching Local Minimum
# https://codeforces.com/contest/1479/problem/A

"""
Algorithm: Binary Search for Local Minimum
Purpose: Find any index k in a permutation array such that a[k] is a local minimum,
         i.e., a[k] < min(a[k-1], a[k+1]), where out-of-bounds elements are considered +∞.

Techniques:
- Binary search on the array indices to efficiently locate a local minimum.
- At each step, query two adjacent elements and compare them to decide which half to explore next.

Time Complexity: O(log n) - Binary search over n elements.
Space Complexity: O(1) - Only using a constant amount of extra space (variables l, r, m).

Note: This approach relies on the fact that in any permutation, there must exist at least one local minimum.
The key insight is that we can use the comparison of adjacent elements to determine which side of the array
to continue searching for a local minimum.
"""

n = int(input())
t = [0] * n
l, r = 0, n
while r - l > 1:
    m = (l + r) // 2
    print("?", m)
    x = int(input())
    print("?", m + 1)
    y = int(input())
    if x < y:  # If left element is smaller than right, a local minimum lies to the left (including m)
        r = m
    else:      # Else, the local minimum lies to the right (including m + 1)
        l = m
print("!", r)  # r now points to a local minimum index


# https://github.com/VaHiX/CodeForces/